client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryRepeatedMappingTest.java

```diff
@@ -0,0 +1,100 @@
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.Search;
+import org.infinispan.client.hotrod.test.HotRodClientTestingUtil;
+import org.infinispan.client.hotrod.test.SingleHotRodServerTest;
+import org.infinispan.commons.configuration.XMLStringConfiguration;
+import org.infinispan.commons.dataconversion.internal.Json;
+import org.infinispan.commons.util.Util;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.query.dsl.Query;
+import org.infinispan.server.core.admin.embeddedserver.EmbeddedServerAdminOperationHandler;
+import org.infinispan.server.hotrod.HotRodServer;
+import org.infinispan.server.hotrod.configuration.HotRodServerConfigurationBuilder;
+import org.testng.annotations.Test;
+
+/**
+ * Tests protobuf mapping with indexed fields and non-indexed repeated field.
+ *
+ * @since 12.1
+ */
+@Test(testName = "client.hotrod.query.RemoteQueryRepeatedMappingTest", groups = "functional")
+public class RemoteQueryRepeatedMappingTest extends SingleHotRodServerTest {
+   private static final String CACHE_NAME = RemoteQueryRepeatedMappingTest.class.getName();
+   private static final String SCHEMA_FILE = "indexed-repeated.proto";
+
+   @Override
+   protected HotRodServer createHotRodServer() {
+      HotRodServerConfigurationBuilder serverBuilder = new HotRodServerConfigurationBuilder();
+      serverBuilder.adminOperationsHandler(new EmbeddedServerAdminOperationHandler());
+      return HotRodClientTestingUtil.startHotRodServer(cacheManager, serverBuilder);
+   }
+
+   @Override
+   protected RemoteCacheManager getRemoteCacheManager() {
+      return super.getRemoteCacheManager();
+   }
+
+   @Test
+   public void testCreateAndQuery() throws Exception {
+      registerProtoBuf();
+
+      RemoteCache<Object, Object> cache = remoteCacheManager.administration().withFlags(VOLATILE)
+            .createCache(CACHE_NAME, createCacheXMLConfig());
+
+      DataFormat dataFormat = DataFormat.builder().keyType(APPLICATION_JSON).valueType(APPLICATION_JSON).build();
+
+      RemoteCache<byte[], byte[]> jsonCache = cache.withDataFormat(dataFormat);
+
+      jsonCache.put(keyAsJson(), valueAsJson());
+
+      Query<Object> querySlowChildren = Search.getQueryFactory(cache).create("SELECT COUNT(*) FROM Parent p WHERE p.slowChildren.id = 0");
+      Query<Object> queryFastChildren = Search.getQueryFactory(cache).create("SELECT COUNT(*) FROM Parent p WHERE p.fastChildren.id = 10");
+      Query<Object> queryFieldChildren = Search.getQueryFactory(cache).create("SELECT COUNT(*) FROM Parent p WHERE p.fieldLessChildren.id = 0");
+
+      assertEquals(1, querySlowChildren.execute().hitCount().orElse(-1));
+      assertEquals(1, queryFastChildren.execute().hitCount().orElse(-1));
+      assertEquals(1, queryFieldChildren.execute().hitCount().orElse(-1));
+   }
+
+   private byte[] keyAsJson() {
+      return Json.object().set("_type", "int32").set("_value", "1").toString().getBytes(UTF_8);
```

Comment:
protobuf <-> json mapping requires the _type for primitives and objects. Json.make(1) would result in ```1```` being sent and no way to indentify the data type from the server POV

Reply:
Ok, I didn't realize we need the `_type` for json->protostream transcoding.

Makes me think though: we say we have json <-> protostream transcoding, but the JSON format we support isn't really JSON, because we require the `_type` attribute. Is this documented? And how hard would it be to support JSON without the `_type` attribute?

One could argue there is no single "protostream format", because it identifies message types with number ids instead of message names, and it's really easy for two applications to use the same id for different messages (even two applications in the same organization). I'm not sure if `MessageWrapper` adding new fields or changing its id is in the same category, but it's worth discussing it in the dogfooding call.

Reply:
Suppose you have:
```proto
Message A {
   optional string name = 1;
}

Message B {
  optional string name = 1;
}
```

Then the cache receives the document:

```json
{
  "name": "I am a JSON"
}
````

That is the rationale for the metafield  ```_type```

Reply:
>  isn't really JSON

Don't agree :smile:

> And how hard would it be to support JSON without the _type attribute?

No idea how to do it. You'd probably have to introduce single ```Message``` caches of some sort, but this would break query where it supports multiple entities (Message) per cache. I don't see any reason to pursue this, what is your concern?

> Is this documented? 

There isn't a dedicated section for it, but all the samples use the _type field.

Reply:
@gustavonalle I understand the rationale behind the `_type` field for complex messages (the ones that have their own Protobuf message name or Protostream proto-id).

But this example does not use an application message: it uses the native Protobuf support for primitive types. And if it looks like a "naked" int in Protostream then it should look like a native int in JSON as well IMO.

Going further, once Adrian's support for collections in `WrappedMessage` lands in Infinispan, then I would expect a JSON array to be mapped as a `WrappedMessage`-based collection in Protostream. Later, when we'll have map support as well, I would expect a structured JSON object without any `_type` field to be mapped to a combination of primitive/collection/map `WrappedMessage`s in Protostream.

Reply:
sorry @danberindei for some reason I did not catch the angle you were coming from :)

Regarding primitives, yes, it'd be nice to have it as a literal JSON, but AFAICR, I had trouble mapping it correctly, since JSON has a single ```number``` type and protobf avout a dozen numeric types that are encoded differently

Reply:
I created https://issues.redhat.com/browse/ISPN-12772 for this